File: 3_Codes/vfi_deterministic.m
% Macro III
% Canonical Growth Model with No Labour and No Investment Constraint
% DETERMINISTIC MODEL
% Equations:
% V(k) = max_{c,k} ln(c) + beta*V(k')
% where
% c + k' = k^(alpha) + (1-delta)*k
% c>=0 and k'>=0
% SOLUTION METHOD: VALUE FUNCTION ITERATION (Discretisation) and Howards
% Improvement algorithm
% Diego de Sousa Rodrigues
% Fall, 2021

clear; clc;

%% Define Preliminaries

% Parameters
alpha = 0.36;   % Capital Share in Production Function
beta  = 0.99;   % Discount Factor
delta = 0.025;  % Depreciation Rate

% Utility Function 
u = @(ct) log(ct);

% Production Function 
F = @(kt) kt^(alpha);

% Define the Steady State
k_ss = (alpha/((1/beta) - (1-delta)))^(1/(1-alpha));  % Steady State Capital Stock
y_ss = F(k_ss);                     % Steady State Output
c_ss = y_ss - delta*k_ss;           % Steady State Consumption

%% Define the Grid for the Endogenous State Variable: Capital
nk = 501;                            % Number of Grid Points
kmin = 0.2*k_ss; kmax = 1.8*k_ss;     % Bounds for Grid
kg = [kmin:(kmax-kmin)/(nk-1):kmax];  % Equally Spaced Grid for Capital

%% Build the 2-Dimensional Contemporaneous Utility Grid
% Within it, ensure that capital and consumption never go negative

Ut = zeros(nk,nk);

for ii = 1:nk       % Loop Over Capital Today
    for jj = 1:nk   % Loop Over Capital Tomorrow
        k  = kg(ii);
        kp = kg(jj);
        % Solve for Consumption at Each Point
        c = F(k) + (1-delta)*k - kp;
        if (kp < 0)||(c < 0)   
            % If Tomorrow's Capital Stock or Today's Consumption is Negative
            Ut(ii,jj) = -9999999999; 
            % Numerical Trick to ensure that the Value Function is never 
            %    optimised at these points
        else
            Ut(ii,jj) = u(c);
        end
    end
end

%% Define an Initial Guess for the Value Function
V0 = ones(nk,1);    % nk x 1 vector of initial guess
V1 = zeros(nk,1);   % nk x 1 vector for optimal value function for given position on capital grid
Objgrid = zeros(nk,nk);  % nk x nk matrix where rows denote today's capital and columns tomorrow's

% Value Function Iteration
tol  = 0.0001;
err  = 2;
iter = 0;

while err > tol
    for ii = 1:nk       % Loop over today's capital stock
        for jj = 1:nk   % Loop over tomorrow's capital stock
            Objgrid(ii,jj) = Ut(ii,jj) + beta*V0(jj);
        end
        [V1(ii,1),PF(ii,1)] = max(Objgrid(ii,:));
    end
    for hh=1:20 % Howards' improvement algorithm
        for ii=1:nk,
            if hh==1
                V11(ii,1)=Ut(ii,PF(ii,1))+beta*V1(PF(ii,1),1);
            elseif hh>1
                V11(ii,1)=Ut(ii,PF(ii,1))+beta*V11(PF(ii,1),1);
            end
        end
    end
    V1=V11;
    iter = iter + 1;
    err  = norm(V1(:) - V0(:));
    iter10 = mod(iter,10);
    if iter10 == 0
        display(['Number of Iterations ',num2str(iter)]);
    end
    V0   = V1;
end

%% Build Policy Function for Consumption
CF = zeros(size(PF));

for j = 1:nk
	k  = kg(j);         % Capital Today
    kp = kg(PF(j,1));   % Capital Tomorrow
    
    % Solve for Consumption
    CF(j,1) = F(k) + (1-delta)*k - kp;
end

%% Plot the Policy Function
fig1 = figure('units','normalized','outerposition',[0 0 0.8 1])
    set(fig1,'Color','white','numbertitle','off','name','Policy Function - Capital')
    plot(kg,kg(PF(:,1)),'k','LineWidth',2)
    hold on
    plot(kg,kg,'k:','LineWidth',1)
    hold off
    xlabel('k Today')
    ylabel('k Tomorrow')
    title('Capital Policy Function')    

File: 3_Codes/vfi_deterministic.jl

##Define Parameters
α =  0.36 # Capital share in production function
β = 0.99 # Discount factor
δ = 0.025 # Depreciation rate


##Utility function
u = (ct) -> log(ct)


##Producution function
F = (kt) -> kt^(α)


##Defining the steady state
k_ss = (α/((1/β) - (1-δ)))^(1/(1-α));  # Steady State Capital Stock
y_ss = F(k_ss);                # Steady State Output
c_ss = y_ss - δ*k_ss;           # Steady State Consumption


##Defining the Grid for the Endogenous State Variable: Capital
nk = 501;                            # Number of Grid Points
kmin = 0.2*k_ss; kmax = 1.8*k_ss;    # Bounds for Grid
kg = kmin:(kmax-kmin)/(nk-1):kmax; # Equally Spaced Grid for Capital


## Build the 2-Dimensional Contemporaneous Utility Grid for the System
Ut = zeros(nk,nk)

for ii = 1:nk       # Loop Over Capital Today
    for jj = 1:nk   # Loop Over Capital Tomorrow
        k = kg[ii];     # Capital Today
        kp = kg[jj];    # Capital Tomorrow
        # Solve for Consumption at Each Point
        c = F(k) + (1-δ)*k - kp
        if (kp < 0)||(c .< 0)
            # If Tomorrow"s Capital Stock | Today"s Consumption is Negative
            Ut[ii,jj] = -9999999999;
            # Numerical Trick to ensure that the Value Function is never
            # optimised at these pointselse
        else
            Ut[ii,jj] = u(c);
            # Calculate Utility at this Point on Grid
        end
    end
end


##Value Function Iteration

#Initial Guess of the Value Function
V0 = ones(nk,1) # nk x 1 vector of initial guess

tol = 1e-10;
its = 0;
maxits = 3000; # Define the maximum number of iterations
V_1 = V0;  # The new value function I obtain after an iteration
V_0 = V0;  # the  value function from which I start in each new iteration
dif = 1;
policy_k_index = Array{Int64,1}(undef,nk);
maxits = 3000


while dif>tol && its <maxits
    global its = its + 1;
    V_0 = copy(V_1);
    for k0 in 1:nk
    V_1[k0], policy_k_index[k0] = findmax(Ut[k0,:] + β*V_0[:])
    end
global dif = maximum(abs.(V_1-V_0));
println("Iteration: $(its)")
end


##Policy function for capital
policy_k = Array{Float64,1}(undef,nk);

policy_k = kg[policy_k_index]


##Policy function for consumption
policy_c = Array{Float64,1}(undef,nk)

for k1 in 1:nk
policy_c[k1] = F(kg[k1]) + (1-δ)*kg[k1] - policy_k[k1]
end


##Plotting the results
using Plots
plot(kg,V_1,title="Value Function",label="Value Function")
plot(kg,policy_c,title="Policy functions for consumption",label="Optimal Consumption")
plot(kg,policy_k,title="Policy functions for capital",label="Optimal Capital")
